File: Stage25/C/Problem1654.c
#include <stdio.h>

int main(void) {
    int K, N;
    scanf("%d %d", &K, &N);

    long long arr[K];
    long long max = 0;

    for (int i = 0; i < K; i++) {
        scanf("%lld", &arr[i]);
        if (arr[i] > max) max = arr[i];
    }

    long long left = 1;
    long long right = max;
    long long answer = 0;

    while (left <= right) {
        long long mid = (left + right) / 2;

        long long count = 0;
        for (int i = 0; i < K; i++) {
            count += arr[i] / mid;
        }

        if (count >= N) {
            answer = mid;
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    printf("%lld", answer);
    return 0;
}

File: Stage25/Python/Problem1654.py
import sys
input = sys.stdin.readline

def main():
    K, N = map(int, input().split())
    arr = []
    
    max_len = 0
    for _ in range(K):
        x = int(input())
        arr.append(x)
        if x > max_len:
            max_len = x
        
    left = 1
    right = max_len
    answer = 0
    
    while left <= right:
        mid = (left + right) // 2
        
        count = 0
        for x in arr:
            count += x // mid
        
        if count >= N:
            answer = mid
            left = mid + 1
        else:
            right = mid - 1
        
    print(answer)
    
if __name__ == "__main__":
    main()